react/prop-types - False positive when ternary conditional return

<!--
    ESLint adheres to the Open JS Foundation Code of Conduct:
    https://eslint.org/conduct

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://eslint.org/chat

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->


**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** 7.5.0
* **ESLint Plugin React Version:** 7.20.3
* **Node Version:** 10.15.3
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
   extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:jsx-a11y/recommended',
    // Prettier plugin and recommended rules
    'prettier/@typescript-eslint',
    'plugin:prettier/recommended',
  ],
  rules: {
    // Include .prettierrc.js rules
    'prettier/prettier': ['error', {}, { usePrettierrc: true }],
    'react/prop-types': 'error',
    'react/react-in-jsx-scope': 'error',
    '@typescript-eslint/explicit-function-return-type': 'error',
    '@typescript-eslint/ban-ts-ignore': 'off',
    'jsx-a11y/label-has-associated-control': [
      'error',
      {
        labelComponents: [],
        labelAttributes: [],
        controlComponents: [],
        assert: 'either',
        depth: 25,
      },
    ],
    'jsx-a11y/anchor-is-valid': 'off',
    '@typescript-eslint/no-explicit-any': 'off',
    '@typescript-eslint/explicit-module-boundary-types': 'error',
    '@typescript-eslint/no-unused-vars': 'error',
    "sort-imports": ["error", {
      "ignoreCase": false,
      "ignoreDeclarationSort": false,
      "ignoreMemberSort": false,
      "memberSyntaxSortOrder": ["none", "all", "single", "multiple"],
      "allowSeparatedGroups": true
    }],
    "@typescript-eslint/ban-types": [
      "error",
      {
        "types": {
          String: {
            message: 'Use string instead',
            fixWith: 'string',
          },
          Boolean: {
            message: 'Use boolean instead',
            fixWith: 'boolean',
          },
          Number: {
            message: 'Use number instead',
            fixWith: 'number',
          },
          Symbol: {
            message: 'Use symbol instead',
            fixWith: 'symbol',
          },

          Function: {
            message: [
              'The `Function` type accepts any function-like value.',
              'It provides no type safety when calling the function, which can be a common source of bugs.',
              'It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.',
              'If you are expecting the function to accept certain arguments, you should explicitly define the function shape.',
            ].join('\n'),
          },

          // object typing
          Object: {
            message: [
              'The `Object` type actually means "any non-nullish value", so it is marginally better than `unknown`.',
              '- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.',
              '- If you want a type meaning "any value", you probably want `unknown` instead.',
            ].join('\n'),
          },
          '{}': {
            message: [
              '`{}` actually means "any non-nullish value".',
              '- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.',
              '- If you want a type meaning "any value", you probably want `unknown` instead.',
            ].join('\n'),
          },
        },
        "extendDefaults": false
      }
    ]
  },
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
import React from 'react';

interface Props {
  item: any;
}

const SomeComponent: React.FC<Props> = ({ item }: Props) => {
  return item ? <></> : <></>;
};

export default SomeComponent;

```

<!-- Paste the command you used to run ESLint: -->
```bash
eslint --fix . --ext .ts,.tsx,.js,.jsx
```

**What did you expect to happen?**

There should not be an error because `item` had been declared with type `any`. 

**What actually happened? Please include the actual, raw output from ESLint.**

```bash
yarn run v1.17.3
$ eslint --fix . --ext .ts,.tsx

/home/gaara/JS/linked-nextapp/components/business/card/public/PublicCard.tsx
  7:40  error  'item' is missing in props validation  react/prop-types

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

**How do you verify the source of problem**

I had tried the following code, which eslint no longer throw error.

```js
import React from 'react';

interface Props {
  item: any;
}

const SomeComponent: React.FC<Props> = ({ item }: Props) => {
  if(item) {
    return <></>
  }
  return <></>
};

export default SomeComponent;
```

```bash
yarn run v1.17.3
$ eslint --fix . --ext .ts,.tsx
Done in 3.09s.
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -22,6 +22,7 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 * [`no-typos`]: fix crash on private methods ([#3043][] @ljharb)
 * [`jsx-no-bind`]: handle local function declarations ([#3048][] @p7g)
 * [`prop-types`], `propTypes`: handle React.* TypeScript types ([#3049][] @vedadeepta)
+* [`prop-types`], `propTypes`: add handling for `FC<Props>`, improve tests ([#3051][] @vedadeepta)
 
 ### Changed
 * [Docs] [`jsx-no-bind`]: updates discussion of refs ([#2998][] @dimitropoulos)
@@ -32,6 +33,7 @@ This change log adheres to standards from [Keep a CHANGELOG](http://keepachangel
 * [Docs] improve instructions for `jsx-runtime` config ([#3052][] @ljharb)
 
 [#3052]: https://github.com/yannickcr/eslint-plugin-react/issues/3052
+[#3051]: https://github.com/yannickcr/eslint-plugin-react/pull/3051
 [#3049]: https://github.com/yannickcr/eslint-plugin-react/pull/3049
 [#3048]: https://github.com/yannickcr/eslint-plugin-react/pull/3048
 [#3043]: https://github.com/yannickcr/eslint-plugin-react/issues/3043

--- a/lib/util/propTypes.js
+++ b/lib/util/propTypes.js
@@ -100,7 +100,8 @@ module.exports = function propTypesInstructions(context, components, utils) {
   const defaults = {customValidators: []};
   const configuration = Object.assign({}, defaults, context.options[0] || {});
   const customValidators = configuration.customValidators;
-  const allowedGenericTypes = ['React.SFC', 'React.StatelessComponent', 'React.FunctionComponent', 'React.FC'];
+  const allowedGenericTypes = new Set(['SFC', 'StatelessComponent', 'FunctionComponent', 'FC']);
+  const genericReactTypesImport = new Set();
 
   /**
    * Returns the full scope.
@@ -548,7 +549,8 @@ module.exports = function propTypesInstructions(context, components, utils) {
       let typeName;
       if (astUtil.isTSTypeReference(node)) {
         typeName = node.typeName.name;
-        if (!typeName && node.typeParameters && node.typeParameters.length !== 0) {
+        const shouldTraverseTypeParams = !typeName || genericReactTypesImport.has(typeName);
+        if (shouldTraverseTypeParams && node.typeParameters && node.typeParameters.length !== 0) {
           const nextNode = node.typeParameters.params[0];
           this.visitTSNode(nextNode);
           return;
@@ -940,12 +942,19 @@ module.exports = function propTypesInstructions(context, components, utils) {
         return;
       }
 
-      const typeName = context.getSourceCode().getText(annotation.typeName).replace(/ /g, '');
+      if (annotation.typeName.name) { // if FC<Props>
+        const typeName = annotation.typeName.name;
+        if (!genericReactTypesImport.has(typeName)) {
+          return;
+        }
+      } else if (annotation.typeName.right.name) { // if React.FC<Props>
+        const right = annotation.typeName.right.name;
+        const left = annotation.typeName.left.name;
 
-      if (allowedGenericTypes.indexOf(typeName) === -1) {
-        return;
+        if (!genericReactTypesImport.has(left) || !allowedGenericTypes.has(right)) {
+          return;
+        }
       }
-
       markPropTypesAsDeclared(node, resolveTypeAnnotation(siblingIdentifier));
     }
   }
@@ -1033,6 +1042,27 @@ module.exports = function propTypesInstructions(context, components, utils) {
     FunctionExpression(node) {
       if (node.parent.type !== 'MethodDefinition') {
         markAnnotatedFunctionArgumentsAsDeclared(node);
+      }
+    },
+
+    ImportDeclaration(node) {
+      // parse `import ... from 'react`
+      if (node.source.value === 'react') {
+        node.specifiers.forEach((specifier) => {
+          if (
+            // handles import * as X from 'react'
+            specifier.type === 'ImportNamespaceSpecifier'
+            // handles import React from 'react'
+            || specifier.type === 'ImportDefaultSpecifier'
+          ) {
+            genericReactTypesImport.add(specifier.local.name);
+          }
+
+          // handles import { FC } from 'react' or import { FC as X } from 'react'
+          if (specifier.type === 'ImportSpecifier' && allowedGenericTypes.has(specifier.imported.name)) {
+            genericReactTypesImport.add(specifier.local.name);
+          }
+        });
       }
     },
 

--- a/tests/lib/rules/prop-types.js
+++ b/tests/lib/rules/prop-types.js
@@ -3150,6 +3150,8 @@ ruleTester.run('prop-types', rule, {
       },
       {
         code: `
+          import React from 'react';
+
           interface PersonProps {
               username: string;
           }
@@ -3161,6 +3163,8 @@ ruleTester.run('prop-types', rule, {
       },
       {
         code: `
+          import React from 'react';
+
           const Person: React.FunctionComponent<PersonProps> = (props): React.ReactElement => (
               <div>{props.username}</div>
           );
@@ -3169,6 +3173,8 @@ ruleTester.run('prop-types', rule, {
       },
       {
         code: `
+          import React from 'react';
+
           interface PersonProps {
               username: string;
           }
@@ -3180,6 +3186,7 @@ ruleTester.run('prop-types', rule, {
       },
       {
         code: `
+          import React from 'react';
           const Person: React.FunctionComponent<{ username: string }> = (props): React.ReactElement => (
               <div>{props.username}</div>
           );
@@ -3188,12 +3195,89 @@ ruleTester.run('prop-types', rule, {
       },
       {
         code: `
+          import React from 'react';
           type PersonProps = {
               username: string;
           }
           const Person: React.FunctionComponent<PersonProps> = (props): React.ReactElement => (
               <div>{props.username}</div>
           );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import { FunctionComponent } from 'react';
+
+          type PersonProps = {
+              username: string;
+          }
+          const Person: FunctionComponent<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import { FC } from 'react';
+          type PersonProps = {
+              username: string;
+          }
+          const Person: FC<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import type { FC } from 'react';
+          type PersonProps = {
+              username: string;
+          }
+          const Person: FC<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import { FC as X } from 'react';
+          interface PersonProps {
+              username: string;
+          }
+          const Person: X<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import * as X from 'react';
+          interface PersonProps {
+              username: string;
+          }
+          const Person: X.FC<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        // issue: https://github.com/yannickcr/eslint-plugin-react/issues/2786
+        code: `
+          import React from 'react';
+
+          interface Props {
+            item: any;
+          }
+
+          const SomeComponent: React.FC<Props> = ({ item }: Props) => {
+            return item ? <></> : <></>;
+          };
         `,
         parser: parsers['@TYPESCRIPT_ESLINT']
       }
@@ -6703,6 +6787,58 @@ ruleTester.run('prop-types', rule, {
           {
             messageId: 'missingPropType',
             data: {name: 'test'}
+          }
+        ],
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          import React from 'react';
+          interface PersonProps {
+              username: string;
+          }
+          const Person: FunctionComponent<PersonProps> = (props): React.ReactElement => (
+              <div>{props.test}</div>
+          );
+        `,
+        errors: [
+          {
+            messageId: 'missingPropType',
+            data: {name: 'test'}
+          }
+        ],
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          interface PersonProps {
+              username: string;
+          }
+          const Person: X.FC<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        errors: [
+          {
+            messageId: 'missingPropType',
+            data: {name: 'username'}
+          }
+        ],
+        parser: parsers['@TYPESCRIPT_ESLINT']
+      },
+      {
+        code: `
+          interface PersonProps {
+              username: string;
+          }
+          const Person: FC<PersonProps> = (props): React.ReactElement => (
+              <div>{props.username}</div>
+          );
+        `,
+        errors: [
+          {
+            messageId: 'missingPropType',
+            data: {name: 'username'}
           }
         ],
         parser: parsers['@TYPESCRIPT_ESLINT']

